pandas/_libs/index.pyx

```diff
@@ -275,7 +275,14 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
         values = np.array(self._get_index_values(), copy=False)
-        stargets = set(targets)
+        values_mask = isnaobj(values)
+        targets_mask = isnaobj(targets)
+        if targets_mask.any():
+            targets_no_nan = targets.copy()
```

Comment:
is it necessary to copy? could you do `stargets=set(targets)` and `stargets.add(-1)`

Reply:
good call. I can add to set instead